src/main/java/org/alfresco/repo/search/impl/querymodel/impl/db/DBQueryEngine.java

```diff
@@ -205,47 +267,173 @@ public QueryEngineResults executeQuery(Query query, QueryOptions options, Functi
+                {
+                    context.stop();
+                    return;
+                }
+                
+                Node node = context.getResultObject();
+                
+                boolean shouldCache = nodes.size() >= options.getSkipCount();
+                if(shouldCache)
+                {
+                    logger.debug("- selected node "+nodes.size()+": "+node.getUuid()+" "+node.getId());
+                    nodesCache.setValue(node.getId(), node);
+                }
+                else
+                {
+                    logger.debug("- skipped node "+nodes.size()+": "+node.getUuid()+" "+node.getId());
+                }
+                
+                if (permissionAssessor.isIncluded(node))
+                {
+                    nodes.add(shouldCache ? node : null);
+                }
+                
+                if (permissionAssessor.shouldQuitChecks())
+                {
+                    context.stop();
+                    return;
+                }
+            }
+        });
+
+        int numberFound = nodes.size();
+        nodes.removeAll(Collections.singleton(null));
+        
+        DBResultSet rs =  createResultSet(options, nodes, numberFound);
+        FilteringResultSet frs = new FilteringResultSet(rs, formInclusionMask(nodes));
+        frs.setResultSetMetaData(new SimpleResultSetMetaData(LimitBy.UNLIMITED, PermissionEvaluationMode.EAGER, rs.getResultSetMetaData().getSearchParameters()));
+ 
+        logger.debug("- query is completed, "+nodes.size()+" nodes loaded");
+        return frs;
+    }
+
+    private DBResultSet createResultSet(QueryOptions options, List<Node> nodes, int numberFound)
+    {
+        DBResultSet dbResultSet = new DBResultSet(options.getAsSearchParmeters(), nodes, nodeDAO, nodeService, tenantService, Integer.MAX_VALUE);
+        dbResultSet.setNumberFound(numberFound);
+        return dbResultSet;
+    }
+
+    private int computeRequiredNodesCount(QueryOptions options)
+    {
+        if (options.getMaxItems() == -1)
+        {
+            return Integer.MAX_VALUE;
         }
-        List<Long> nodeIds = new ArrayList<Long>(set);
-        ResultSet rs = new DBResultSet(options.getAsSearchParmeters(), nodeIds, nodeDAO, nodeService, tenantService, Integer.MAX_VALUE);
-        ResultSet paged = new PagingLuceneResultSet(rs, options.getAsSearchParmeters(), nodeService);
+        
+        return options.getMaxItems() + options.getSkipCount() + 1;
+    }
 
-        Map<Set<String>, ResultSet> answer = new HashMap<Set<String>, ResultSet>();
+    private BitSet formInclusionMask(List<Node> nodes)
+    {
+        BitSet inclusionMask = new BitSet(nodes.size());
+        for (int i=0; i < nodes.size(); i++)
+        {
+            inclusionMask.set(i, true);
+        }
+        return inclusionMask;
+    }
+
+    
+    private QueryEngineResults asQueryEngineResults(ResultSet paged)
+    {
         HashSet<String> key = new HashSet<String>();

```

Comment:
Can just use inference <> here

Reply:
Yes I guess we can, that was part of the code that we found in the class and we just extracted in a separate method.